### ISSUE TITLE
I need a "spinning phasor" object to take FM and convert to PM.

### ISSUE THREAD

**Issue by elanhickler** [RobinSchmidt/RS-MET#106]
Hey Robin, I just got this idea to solve a problem with modulating things with anti-aliased phasors (or sawtooths/ramps).

I need to use the rotation formula to rotate a signal in two or three dimensions, for example: 

```
void rotate2D(double sin, double cos, double * X, double * Y)
{
	double x = *X, y = *Y;
	*X = x * cos - y * sin;
	*Y = x * sin + y * cos;
}
```

So I need to derive a sin/cos out of a velocity of change of value rather than a fixed value/DC/phase value.

Imagine taking your hand and spinning a floating wheel. Now try to stop the wheel from spinning as fast as you can and make it spin the other direction. The speed of the movement of your hand hand represents frequency, and your strength represents ... force? torque? Now replace your hand with a sawtooth or a phasor. All we need to do is have a parameter for "how fast can the sawtooth stop/reverse the spinning of the wheel".

I'm trying to solve this problem:

![perfect_vs_antialiased_saw](https://user-images.githubusercontent.com/9116575/31788729-8745941e-b4c4-11e7-89d4-e32c4340a662.png)

My LFO is an anti-aliased saw and I want to control rotation. But the glitch in modulation of the "smoothed slope" causes the sound to click before rotation is set back to 0. If the anti-aliased saw could control a factor of rotation RATHER THAN the rotation value direction, the rotation glitch could be smoothed away.

![image](https://user-images.githubusercontent.com/9116575/31788999-7322d3f6-b4c5-11e7-97e7-6a443eae5187.png)

**Comment by RobinSchmidt** [comment]
i'm not sure if i get this right, but it almost sounds to me like you are talking about just smoothing the edge here. what's the difference to using a simple lowpass filter? i think, your graph with the spike (if that represents a force, hence an acceleration) and the smoothed out version of it, is what a lowpass would do. btw. integrating periodic bipolar spikes gives a sawtooth wave.

needless to say, however, that smoothing of course doesn't really anti-alias anything unless you use it together with oversampling

**Comment by RobinSchmidt** [comment]
and what have sawtooth waves to do with rotations? they create circular motion (i.e. sines and cosines)

**Comment by elanhickler** [comment]
robin, a lowpass filter is the PROBLEM! The saw is lowpassed... it's anti-aliased. That being said, do you get it now or should I make some more graphs? 😃

**Comment by RobinSchmidt** [comment]
wait wait wait...so you are using a sawtooth "phasor" for the phase-input of a sine wave, i.e. a sine-function, right (or maybe even some other wavshape)? when the phasor hits 1, it jumps back to 0 (or -1 whatever your conventions). ....and you picture this as the phasor turning around and going back in a very short amount of time? i don't really think that this is the right mental model for what the phasor does. you can just as well imagine it increasing forever. in that case, the periodicity of the sin-function would take over the role of resetting the phasor. the phasor just passes 1 ...but that value > 1 is the equivalent to some other value (shortly above 0). there is no turning back and then picking up forward speed again

**Comment by RobinSchmidt** [comment]
it's like you are imagining a rotating wheel that can only rotate once and then has to be rewinded back in order to allow for another forward rotation. ...but it actually just continues to rotate forward

**Comment by RobinSchmidt** [comment]
as far as the sine is concerned 0.1 is the same input as 1.1, the 1.1 is just mapped back to 0.1. it doesn't need to *go* back by reversing direction

**Comment by elanhickler** [comment]
> it's like you are imagining a rotating wheel that can only rotate once and then has to be rewinded back in order to allow for another forward rotation. ...but it actually just continues to rotate forward
yes almost! it's better to imagine replacing the sawtooth with your hand. You HAVE to move your hand back up to push the wheel forward again. But then your hand must stay in contact with the wheel, and essentially "slip across" rather than your hand being an infinitely strong force that stops the wheel from spinning... or... something. Wow this kind of sounds like soft sync. The wheel could only turn / speed up /slow down efficiently if the speed of your hand is around the speed of the wheel. Go too fast and your hand slips off and doesn't increase the speed much. Go too slow and you just have a low speed.

....anyway, so would your suggestion be doing something where I only use a portion of the sawtooth (the non-glitched part) somehow?

**Comment by RobinSchmidt** [comment]
i'm still totally confused. there is no "glitch" in the sawtooth and i don't know how to imagine to "replace the sawtooth" with my hand. the sawtooth shape is actually just an artifact of our desire to avoid numerical overflow - otherwise it would just be an ever increasing linear function. i think, it's better to replace the sawtooth with that fucntion in your mental model. that would be more accurate. the sawtooth value does not "jump back" in the sense of moving back at infinite speed. it just disappears at the top and reappears at the bottom. it's actually wrong to connect the two values with a vertical line

**Comment by elanhickler** [comment]
the sawtooth shape is *not* an artifact of a desire to avoid numerical flow, the sawtooth shape is the correct shape of a physically-limited saw within the increment of time of the digital computer that is simulating reality in 1/44100 steps of time.

But forget all that for a moment, the imperfect sawtooth shape produces a more pleasing sound because it is warmer and more physically limited/accurate for the limitation of time steps of the digital world.

I think you just have to experience the problem for yourself. Modulate stereo rotation in jerobeam mushroom with my sawtooth LFO. Then think about how to solve the issue.

Or forget everything I said for a moment and focus on this one problem: Convert FM to PM. Take a frequency modulation signal and turn that into a phase modulation value. This is a problem we should solve if we are going to use oscillators that don't have a phase input.

**Comment by RobinSchmidt** [comment]
> focus on this one problem: Convert FM to PM

the instantaneous phase of a signal is given by the integral (think: running sum) of the instantaneous frequency. so i guess, going back and forth between FM and PM involves integrating or differentiating the modulator signal (i.e. taking a running sum y[n] = y[n-1] + x[n] or a 1st difference y[n] = x[n] - x[n-1]). ...may need some more attention to detail, though

**Comment by elanhickler** [comment]
ok good! You have the problem in your head! That's all I need. I'll pay for your time working on this when I really need it solved. I'm going to release JB mushroom now and just allow the user to select non-aa saw.

**Comment by elanhickler** [comment]
just in case, maybe a diagram will help:
![image](https://user-images.githubusercontent.com/9116575/31798887-7386a90a-b4eb-11e7-820d-fde61d90b26b.png)

**Comment by RobinSchmidt** [comment]
i think, we are on the right track with integration/differentiation:

https://moinsound.wordpress.com/2011/03/04/frequency-modulation-or-phase-modulation-synthesizer-technologies/

which boils down to 1st lowpass/highpass filters. a running summ is basically a 1st order lowpass with 0 cutoff frequency and a differentiator a highpass at fs/2. however, i can imagine that the modulator needs some rescaling after filtering which should somehow track the modulator frequency

**Comment by elanhickler** [comment]
Hey Robin, remember this problem? I want you to work on it! FM to PM converter. I want to experiment with this as a way to put "physical limitations" on modulations, so we can have smoother modulation that doesn't break the sample rate limit causing harsh tones.

**Comment by elanhickler** [comment]
Oh, I remember what I wanted to do with this. I wanted to create a sinusoid generator which only has frequency as a parameter. It's also not reliably in tune.  What's that sinusoid equation?

something like:

```cpp
s0 -= f * s1;
s1 += f * s0;
```

So it only has frequency and I need a PM to FM converter to do PM. AND it doesn't have to be perfect, I'm purposely interested in it not being perfect, and maybe we need a parameter for how imperfect, or how quickly frequency is allowed to change or something. I don't know what your end result will be so I don't know what parameters you will have.

**Comment by RobinSchmidt** [comment]
what is this update equation?
```
s0 -= f * s1;
s1 += f * s0;
```
is this supposed to be a recursive sine oscillator? this would have to be the cheapest update equation i have ever seen for such a thing. are you sure, this works? were does it come from?

**Comment by elanhickler** [comment]
Man it would be nice if you could get your modular synth working so we could do all these experiments quickly 😄

**Comment by elanhickler** [comment]
So your mission is to create a kind of phase modulation version of that sine update equation with some kind of PM to FM conversion.

**Comment by RobinSchmidt** [comment]
>  it would be nice if you could get your modular synth working

yeahh. i just re-organized the modules in ToolChain, making an "Under Construction" section at the bottom where i have now included Liberty. somehow i don't get sound out of it and the colors look weird but maybe i can get it to work soon. the main challenge was to get the whole software architecture working - but this is done and what remains is adding more atomic modules - a straightforward task of wrapping some bits of core dsp code into the liberty api, so to speak - but somehow i then didn't bother to add any more modules...

this simple recursive sine equation is interesting. i must experiment with it. my currently simplest sine oscillator is in class rsSineIterator:
```cpp
  RS_INLINE T getValue()
  {
    T tmp = a1*s1 - s2;
    s2 = s1;
    s1 = tmp;
    return tmp;
  }
```
...but this one is perfect

**Comment by RobinSchmidt** [comment]
btw - the filter i was recently talking about (with the supposedly perfect modulation response) is based on the rotation-matrix idea, too. but with a few tweaks to allow for non-resonant filters as well (some filters cannot be expressed as rotation matrix - but they can use another kind of 2x2 matrix)

**Comment by RobinSchmidt** [comment]
> So your mission is to create a kind of phase modulation version of that sine update equation with some kind of PM to FM conversion.
> 
> (Edit: And I want to use that PM to FM conversion in other places, so isolate it as its own thing)

i think, if only sine-waves are assumed, the task will be simpler than in the general case - but i also think, it could be interesting to solve the problem for the general case and then make a simplified version for the sine case. the thing is the integration/differentiation stuff is actually some sort of lowpass/highpass filter (possibly with some gain-compensation involved) in the general case whereas it might be simple first order allpass in the sine case (the derivative of a sine is a cosine and you can get that with a properly tuned allpass)

**Comment by RobinSchmidt** [comment]
yes, i mean if the modulation signal (i.e. the modulator) is as sine wave. in this case, the conversion by differentiation (to cosine) is a simple allpass and likewise, the integration to negative cosine is the same allpass with sign inversion. otherwise, integrators and differentiators are perhaps a little more complicated - but probably not really very much - it would be just  lowpass and highpass - but perhaps  with some gain compensation which might become a bit tricky...but maybe not...we'll see

**Comment by RobinSchmidt** [comment]
i'm currently making my wavetable-oscillator from straightliner available as source module in ToolChain. then i will combine 2 of them into a dual-oscillator with interactions - FM/PM/sync/etc. - then i have a good environment to experiment with FM

**Comment by RobinSchmidt** [comment]
> i think, liberty would actually also be a good environment to experiment with FM vs PM

i will try to get liberty ready for work and then provide a sine-osc module with phase input and then create a PM patch and then try to emulate it using only the frequency input and filters. i think, passing the modulator signal through a differentiator before entering the frequency input should do the trick. likewise, using an integrator on the modulator in front of the PM input should emulate FM via PM

**Comment by elanhickler** [comment]
Hmm, if you get Liberty working now I might actually setup Chaosfly in it and just go without having signal visualizers. Oh! Here's a solution, have many outputs out of Liberty and connect those outputs to VST audio visualizers!

**Comment by elanhickler** [comment]
And I need to know how to create modules for Liberty, I need a bunch of custom ones. Maybe I'll have to keep a personal repository for all my Liberty modules.

**Comment by RobinSchmidt** [comment]
> have many outputs out of Liberty and connect those outputs to VST audio visualizers!

for that, i would first have to figure out, how to let ToolChain itself have a dynamic number of outputs. something, that's on the ToDo list because i want to make my Linkwitz/Riley crossover available in ToolChain - and possibly the new crossover from multicomp (which itself will also have to be improved).

custom (atomic) modules...phew...i'll to refresh my grasp of the api that i write back then (fortunately, i took the code documentation seriously). from a quick look, it seems you need to subclass from `romos::ModuleAtomic` but there are some things about the api that i'd like to change - for example, i have been using my own string class - i'd like to replace that with std::string. what kind of modules do you need? maybe they can themselves be realized meaningfully in a non-atomic way, given the right atomic modules

**Comment by RobinSchmidt** [comment]
for example, i'm planning a tapped delay-module that has outputs for x[n-1], x[n-2],.... - a basic building block for filters. ...and a formula module to allow arbitrary N-input-M-output formulas (re-using the expression evaluator from func shaper)

**Comment by elanhickler** [comment]
Maybe I can ask you to add certain building blocks. Example would be the trisaw oscillator or resonator filters, signal crossfader for FM/PM.

**Comment by RobinSchmidt** [comment]
i think, i will make a TriSaw shaper into which you can feed input from a phasor. this is more flexible

**Comment by elanhickler** [comment]
That's fine, it's less convenient, so we will need a feature where you create a container in Liberty, and you can duplicate that container, and a change to one of those containers updates the rest of the containers. "Macro Container"?

**Comment by RobinSchmidt** [comment]
you can already create containers. either by putting a fresh container into the structure and going into it or by selecting a bunch of modules and "containerizing" the selection (all in the right-click menu). when you click on a container from it's parent module, you also have a "save container" option and when you right-click on an empty area, you can load a container....but yeah - direct duplication would be more convenient

**Comment by elanhickler** [comment]
but if you change the container, only that one container changes? The container should reference the saved container file so that all containers referencing that file will be updated to reflect the changes.

**Comment by RobinSchmidt** [comment]
ahh...now i understand what you mean some kind of aliasing or linking of a bunch of containers. but what has that to do with whether we have a TriSawOsc or TriSawShaper. the TriSawOsc is supposed to created from the TriSawShaper and a Phasor just once and for all and then used in the same, unchanged form everywhere - in the same way an atomic module would (this is why i didn't get what you mean at first)

**Comment by elanhickler** [comment]
yes, create it once, but for my purposes I may want to add more features like a convenient parameter for self-modulation. I want to create a whole Chaosfly Oscillator Unit which may have ever-growing number of features to make it easy to experiment with. I might build the soft clipper directly into the oscillator unit.

**Comment by RobinSchmidt** [comment]
i've been just re-reading some document i printed out many years ago. this one:

http://argos.vu/wp-content/uploads/2016/05/Digital_Sound_Generation_1.pdf

on page 40, it says:

> Even if an oscillator lacks immediate phase control, it’s still possible to add FM capabilities.
> As phase is the time integral of frequency, we may differentiate the modulator signal and add
> it to the frequency input

**Comment by elanhickler** [comment]
Move this to Discussion / Ideas
